### PR TITLE
feat: multi currency support

### PR DESCRIPTION
**Pull request by 0xdef1cafe**
## Description

* builds on top of other PR where data fetching is implemented
* adds view layer changes required to consume this data
* adds settings selection switcher behind flag
* **swapping is not implemented in this PR - and will be added in another PR**

## Notice

- [x] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

wip

## Risk

significant - touches pretty much everything.

## Testing

full regression test of the app with the flag *OFF* (default) required

## Screenshots (if applicable)

no visual changes should be present

### CODE REVIEW

**Review by cjthompson — APPROVED** [review]

**Review by gomesalexandre — CHANGES_REQUESTED** [review]
- Still seeing that $0 vs. $0.00 formatting, both with the flag on and off.

image
image

- The wrong separator is used for Euro, which should be spaces as a separator for three-digit groups https://web.archive.org/web/20061127224154/https://www.nist.gov/public_affairs/techbeat/tb2006_1122.htm#decimal. It can either be a comma or a period for the decimal point separator, so no need to format it any special way, although having a comma as a decimal separator would probably feel more natural for Euro. 

Apart from that, tested with the flag on and the right selected currency is applied across the app, and I didn't find any regressions with the flag off except for the $0 mentioned above.

File: src/hooks/useLocaleFormatter/useLocaleFormatter.ts
Comment (chxash) [reply src/hooks/useLocaleFormatter/useLocaleFormatter.ts:47]: @gomesalexandre do we still want to display supply values like that `386.04M` vs `386,046,170` ?
I find it less pretty to display all numbers for such big values.

Comment (gomesalexandre) [reply src/hooks/useLocaleFormatter/useLocaleFormatter.ts:47]: Yeah looks like we'll be displaying the full amount with the standard `<Amount />` formatting, more accurate as well

**Comment by 0xdef1cafe** [comment]
converting to draft until @stackedq is able to look at currency formatting and chart issues

https://discord.com/channels/554694662431178782/984598894883250226

**Comment by stackedq** [comment]
@willyogo definitely, is there any UI design in your mind for number format selection?

**Comment by 0xdef1cafe** [comment]
@shapeshift/product can we please get some love on the above for @stackedq ?

**Comment by 0xdef1cafe** [comment]
update on this PR - the currency formatting issue on charts is fixed, and i can't see any chart regressions with this flag on and USD selected, or with the flag off.

@stackedq is going to implement the currency selection in the settings modal, and we can merge this with the flag off, and tackle multi currency trade support in a separate PR, then get the flag on and feature released.

**Comment by 0xdef1cafe** [comment]
next issue to implement trade support https://github.com/shapeshift/web/issues/1995

File: src/components/Modals/Settings/SettingsList.tsx
Comment (kaladinlight) [inline src/components/Modals/Settings/SettingsList.tsx:32]: typo on this guy

File: src/hooks/useFetchPriceHistories/useFetchPriceHistories.ts
Comment (kaladinlight) [inline src/hooks/useFetchPriceHistories/useFetchPriceHistories.ts:25]: No longer applicable? removal of `JSON.stringify` doesn't cause infinite re-renders?

File: src/components/Modals/Settings/FiatCurrencies.tsx
Comment (kaladinlight) [inline src/components/Modals/Settings/FiatCurrencies.tsx:5]: smallest of comments, not exactly sure about tree-shaking with the explicit path imports vs destructed on top level lodash, but feel like either or would be better

Comment (gomesalexandre) [reply src/components/Modals/Settings/FiatCurrencies.tsx:5]: Yeah `lodash/functionName` should always be preferred unless you have some AST magic to automate that e.g babel plugin https://lodash.com/per-method-packages

**Review by gomesalexandre — COMMENTED** [review]
Tested both with the flag on and off:

- With the flag on, I couldn't find any place where the selected currency wasn't displayed.
- Seeing the right separator for Euro when selecting French as a language which seems to be the intent from looking at the implementation, but I'm not sure is right. Such separator should be determined by the currency, not the language. Happy for this to be improved in a follow-up PR. 
- The currency is always present before the value, which should be after by convention in the case of e.g Euro. I'm not sure if this is something we want to be granular with depending on the selected currency, but if so, it should be matching the convention of said currency. Again, happy for this to be tackled in a follow-up PR

- I didn't notice any regressions on the existing behavior with the flag off, so I'm happy to merge this as is after the merge conflicts and the blocking comments are tackled.

File: src/hooks/useFetchPriceHistories/useFetchPriceHistories.ts
Comment (0xdef1cafe) [reply src/hooks/useFetchPriceHistories/useFetchPriceHistories.ts:25]: assetIds are now stable coming from the selector so no, `JSON.stringify` would have been a workaround prior

**Comment by gomesalexandre** [comment]
@stackedq Some more conflicts after https://github.com/shapeshift/web/pull/2116 has been merged

**Comment by reallybeard** [comment]
Fixed the conflicts with my new defi ui stuff, a lot of it isn't needed anymore since we left the amount component do the heavy lifting.

I do have a suggestion on the currency selection, can we not sort the list on selection? I'm fine with it being sorted on mount, but its really weird that it jumps the item to the top when you make a selection.

**Comment by stackedq** [comment]
@reallybeard changed some code to achieve the UX you described, please take a look to see if you're happy with the results.

File: src/assets/translations/en/main.json
Comment (gomesalexandre) [inline src/assets/translations/en/main.json:698]: We'll need to get translations for this after this is merged. Kind of weird for a multi-currency / currency format option not to be translated, as it would logically be used by users inclined to selecting another language.

image

File: src/components/AssetHeader/AssetMarketData.tsx
Comment (gomesalexandre) [inline src/components/AssetHeader/AssetMarketData.tsx:110]: Just a syntactic nit but I find the shorthand true syntax to be more declarative

```suggestion
                  abbreviated
                  omitDecimalTrailingZeros
```

(No need to commit the suggestion as it will fail lint and really doesn't matter)

File: src/components/Modals/Settings/SettingsRouter.tsx
Comment (gomesalexandre) [inline src/components/Modals/Settings/SettingsRouter.tsx:23]: Again a nitpick so no need to commit it but we're not using any prop for these two components, so that just makes it one more level in the JSX tree for no reason, this could either be:

```suggestion
        <Route path={SettingsRoutes.FiatCurrencies} component={FiatCurrencies} />
        <Route path={SettingsRoutes.CurrencyFormat} component={CurrencyFormat} />
```

Or use the conventional children syntax for `<Route />`: https://v5.reactrouter.com/web/guides/primary-components/route-matchers

File: src/hooks/useFetchPriceHistories/useFetchPriceHistories.ts
Comment (gomesalexandre) [inline src/hooks/useFetchPriceHistories/useFetchPriceHistories.ts:31]: Would prefer seeing curly braces rather than an obscure casting syntax (also you shouldn't return for a reason, the return value of `useEffect` is actually used and checked for: if it's a function, it will be used as a cleanup function which is why you were seeing type errors)

```suggestion
useEffect(() => {
    dispatch(findPriceHistoryByFiatSymbol.initiate({ symbol, timeframe }))
  }, [dispatch, findPriceHistoryByFiatSymbol, symbol, timeframe]
```

**Review by gomesalexandre — APPROVED** [review]
Retested according to the previous criteria and it generally works as expected (minus a smallish comment WRT translations but that's for globalization and can be done separately) / didn't find any regression.

Re:
> @stackedq I do have a suggestion on the currency selection, can we not sort the list on click? I'm fine with it being sorted on mount, but its really weird that it jumps the item to the top when you make a selection. Same goes for the currency format selection.

Tested this one out and that's behaving as suggested by @reallybeard as well.

Just me being a nit, but I've found it that the second "Currency Format" option isn't exactly matching the example in the option, see the option vs. the actual values formatting, missing the thousands separator:

image

Given this is such a small issue, and this is under feature flag, I'm going to accept this so we can move forward, and tackle small visual issues separately as follow-ups.
My biggest concern given the size of the PR would be whether or not this produces regressions, which from looking at the implementation and testing over, isn't the case.

🍬

**Comment by gomesalexandre** [comment]
Given the high risk of this, let's get @shapeshift/operations to test the Fleek branch

**Comment by reallybeard** [comment]
Looks good @stackedq working really nice!

**Review by gomesalexandre — APPROVED** [review]
Tested again as a sanity check, still working as previously.
The currency format option value is also now matching the actual formatting

🍬

**Review by reallybeard — APPROVED** [review]

File: src/assets/translations/en/main.json
Comment (0xdef1cafe) [reply src/assets/translations/en/main.json:698]: @gomesalexandre this will get picked up by @firebomb1 out of band as he's now owning the translations PRs coming in from gitlocalize

**Comment by 0xdef1cafe** [comment]
going to merge this for the following reasons

* architecturally, this was previously discussed and approved by @cjthompson and recently tested and approved by @gomesalexandre and @reallybeard 
* @shapeshift/operations has signed off on this today
* it's behind a flag
* this branch is a bastard child

**Pull request merged**
